crates/red_knot_python_semantic/resources/mdtest/attributes.md

```diff
@@ -2,6 +2,212 @@
+        self.pure_instance_variable = "value set in method"
+
+c_instance = C()
+
+# for a more realistic example, let's actually call the method
+c_instance.set_instance_variable()
+
+# TODO: should be `str`
+reveal_type(c_instance.pure_instance_variable)  # revealed: @Todo(instance attributes)
+
+# TODO: See above, we currently plan to emit diagnostics for both of these lines,
+# even if mypy/pyright do not.
+reveal_type(C.pure_instance_variable)  # revealed: str
+C.pure_instance_variable = "overwritten on class"
+```
+
+### Pure class variables (`ClassVar`)
+
+#### Annotated with `ClassVar` type qualifier
+
+Class variables annotated with the [`typing.ClassVar`] type qualifier are pure class variables. They
+can not be accessed on instances.
+
+```py
+from typing import ClassVar
+
+class C:
+    pure_class_variable: ClassVar[str] = "value in class body"
+
+reveal_type(C.pure_class_variable)  # revealed: str
+
+c_instance = C()
+
+# TODO: This should be `str`. It is okay to access a pure class variable on an instance.
+reveal_type(c_instance.pure_class_variable)  # revealed: @Todo(instance attributes)
+
+# TODO: should raise an error. It is not allowed to reassign a pure class variable on an instance.
+c_instance.pure_class_variable = "value set on instance"
+
+C.pure_class_variable = "overwritten on class"
+
+# TODO: should ideally be `Literal["overwritten on class"]`, but not a priority
+reveal_type(C.pure_class_variable)  # revealed: str
+
+# TODO: should raise an error (incompatible types in assignment)
+C.pure_class_variable = 1
+```
+
+#### Variable only mentioned in a class method
+
+We also consider a class variable to be a pure class variable if it is only mentioned in a class
+method.
+
+```py
+class C:
+    @classmethod
+    def class_method(cls):
+        cls.pure_class_variable = "value set in class method"
+
+# for a more realistic example, let's actually call the method
+C.class_method()
+
+# TODO: mypy shows an error here, pyright does not. What should we do?
+# error: [unresolved-attribute]
+reveal_type(C.pure_class_variable)  # revealed: Unknown
+
+C.pure_class_variable = "overwritten on class"
+
+# TODO: should be `Literal["overwritten on class"]`
+# error: [unresolved-attribute]
+reveal_type(C.pure_class_variable)  # revealed: Unknown
+
+c_instance = C()
+# TODO: should be `Literal["overwritten on class"]` or `str`
+reveal_type(c_instance.pure_class_variable)  # revealed: @Todo(instance attributes)
+
+# TODO: should raise an error.
+c_instance.pure_class_variable = "value set on instance"
+```
+
+### "Regular" class variables
+
+These are instance attributes, but the fact that we can see that they have a definition (not just a
+declaration) in the class body means that reading the value from the class body is also permitted.
+This is the only difference for these attributes as opposed to "pure" instance attributes.
+
+#### Basic
+
+```py
+class C:
+    regular_class_variable: str = "value in class body"
```

Comment:
same as my comment above about the subheading, I think calling this variable `regular_class_variable` is somewhat confusing, as this category really has more in common with "pure" instance variables than with pure class variables